cunumeric/array.py

```diff
@@ -1991,16 +1994,40 @@ def _diag_helper(self, offset=0, axes=None, extract=True):
 
             tr_shape = tuple(a.shape[i] for i in range(a.ndim - N))
             # calculate shape of the output array
-            out_shape = tr_shape + (diag_size,)
-            out = ndarray(shape=out_shape, dtype=self.dtype, inputs=(self,))
-
-            out._thunk._diag_helper(
-                a._thunk,
-                offset=offset,
-                naxes=N,
-                extract=extract,
+            if trace:
+                if N != 2:
+                    raise ValueError(
+                        " exactly 2 axes should be passed to trace"
+                    )
+                if self.ndim == 2:
+                    out_shape = (1,)
+                elif self.ndim > 2:
+                    out_shape = tr_shape
+                else:
+                    raise ValueError(
+                        "dimension of the array for trace operation:"
+                        " should be >=2"
+                    )
+            else:
+                out_shape = tr_shape + (diag_size,)
+
+            if out is not None:
+                out_arr = convert_to_cunumeric_ndarray(out)
+                if out.shape != out_shape:
+                    raise ValueError("output array has wrong shape")
+                common_type = self.find_common_type(out, a)
+                args = (out, a)
+                a = a._maybe_convert(common_type, args)
+                out = out._maybe_convert(common_type, args)
```

Comment:
This logic is not appropriate here; `find_common_type` is used in the binary operation implementation to find the widest type that will fit both arguments.

However, `out` should never be casted to another type; it is the one setting the dtype of the computation. You need to convert `a` to match `out`'s dtype:

```
a = a._maybe_convert(out_arr.dtype, (a, out_arr))
```